### ISSUE TITLE
Installing python-elasticsearch-curator fails because python-pyyaml-3.11-1.x86_64 conflicts with  PyYAML-3.11-1.el7.x86_64

### ISSUE THREAD

**Issue by martinhynar** [elastic/curator#754]
Hi, I have my nodes managed with salt which is installed from SaltStack's rpm repo. Their repo (http://repo.saltstack.com/yum/redhat/7/x86_64/2015.8/) contains rpm package for PyYAML and salt is dependent on it. After adding curator rpm repo, I am not able to install python-elasticsearch-curator package, because curator repo has (probably the same)  PyYAML named as python-pyyaml package. On installation, this comes to conflict. See the trace of the installation error:

```
  file /usr/lib64/python2.7/site-packages/yaml/__init__.py from install of python-pyyaml-3.11-1.x86_64 conflicts with file from package PyYAML-3.11-1.el7.x86_64
  file /usr/lib64/python2.7/site-packages/yaml/__init__.pyc from install of python-pyyaml-3.11-1.x86_64 conflicts with file from package PyYAML-3.11-1.el7.x86_64
  file /usr/lib64/python2.7/site-packages/yaml/nodes.py from install of python-pyyaml-3.11-1.x86_64 conflicts with file from package PyYAML-3.11-1.el7.x86_64
  file /usr/lib64/python2.7/site-packages/yaml/nodes.pyc from install of python-pyyaml-3.11-1.x86_64 conflicts with file from package PyYAML-3.11-1.el7.x86_64
  file /usr/lib64/python2.7/site-packages/yaml/serializer.py from install of python-pyyaml-3.11-1.x86_64 conflicts with file from package PyYAML-3.11-1.el7.x86_64
...
```

For previous version of curator (targeting ES 1.7.x) this was not a problem as curator was not split into API and CLI parts. Now I am able to only install curator python library but not the curator cli interface. Any hint on how to solve this?

**Comment by untergeek** [comment]
This is RPM hell, as it is colloquially known.  Your options are to:

1. Install the RPM manually, instead of from the curator repo.  You would also need the other dependent RPMs in this case (click, urllib3, elasticsearch, etc.).

2. Install via `pip`, as per the [instructions](https://www.elastic.co/guide/en/elasticsearch/client/curator/current/installation.html) in the documentation.

> Now I am able to only install curator python library but not the curator cli interface. Any hint on how to solve this?

Curator 4 does not separate the API and CLI.  They are the same.  If you can install Curator 4, you have the CLI.  It is just very different from Curator 3 as it uses YAML configuration files instead of a bajillion command-line flags.   Start [here](https://www.elastic.co/guide/en/elasticsearch/client/curator/current/configuration.html) for more information.

**Comment by martinhynar** [comment]
I installed elasticsearch-curator pip package as recommended and started to transform my cron jobs to use action files instead of passing bunch of flags. It is extra work, but from the very beginning I see it is way more flexible and feature rich, so it is worth making.

It is that python-elasticsearch-curator-4.x.x rpm package has those rich command line flags that make its usage handy for ad-hoc checks in terminal.

Anyway, closing this. There is nothing to work on. Maybe, have pyaml as official package, but that different story.

**Comment by untergeek** [comment]
There is a way around this, and that is to compile Curator.

I just built this and verified it against a clean VM with nothing pre-installed.  You don't even necessarily need to have python installed at all, since it's bundled.

# Compile Curator into a binary with included libraries

## Errata

**NOTE:** This how-to is only verified for CentOS 6.  **There are extra steps/workarounds involved to enable cx-Freeze on other platforms.** Also, this build is x86-64 only.  You should be able to modify the steps/VMs to work for other platforms, but that is not covered here.

This requires either `vagrant`, or a CentOS 6 installation with internet access.  These instructions use Vagrant, so you can build it anywhere

### Make a curator build directory (replace `/path/to` with whatever you would like, that you have access to):

```
mkdir /path/to/curator_build
```

### Build a Vagrant VM:

```
mkdir centosVM
cd centosVM
```

### Put the following in a file called Vagrantfile:

```
Vagrant.configure(2) do |config|
  config.vm.box = "elastic/centos-6-x86_64"
  config.vm.synced_folder "/path/to/curator_build", "/curator_build", create: true, owner: "vagrant", group: "vagrant"
  config.vm.provider "virtualbox" do |v|
    v.customize ["modifyvm", :id, "--nictype1", "virtio"]
  end
end
```

Note that `/path/to/curator_build` is in this configuration.  Be sure to replace it with whatever you used in that previous step.

### Run:

```
vagrant up; vagrant ssh
```

### Inside the VM, run: 

```
sudo -i
yum groupinstall "Development Tools" "Development Libraries" python-devel
pip install -U setuptools

wget https://github.com/elastic/curator/archive/v4.1.1.tar.gz 
tar zxf v4.1.1.tar.gz
cd curator-v4.1.1
pip install --user cx-Freeze
pip install --user -r requirements.txt

python setup.py build_exe
mv build/exe.linux-x86_64-2.6 /curator_build/curator-4.1.1
```

### Test the compiled version from this directory:

```
cd /curator_build/curator-4.1.1
./curator --version
```

You should see:
```
curator, version 4.1.1
```

### Exit the vm now:
```
shutdown -h now
```

### Conclusion
When you exit the VM, the fully compiled-for-CentOS 6 version of Curator will be available in `/path/to/curator_build/curator-4.1.1`

You can now tar/zip this directory and copy it to the target machines.  You can also delete the entire VM directory afterward.

**Comment by untergeek** [comment]
I now publish these binaries directly to the APT and YUM repositories:

https://www.elastic.co/guide/en/elasticsearch/client/curator/current/apt-repository.html#apt-binary

https://www.elastic.co/guide/en/elasticsearch/client/curator/current/yum-repository.html#yum-binary
